Libraries/oneMKL/monte_carlo_european_opt/src/montecarlo_main.cpp

```diff
@@ -0,0 +1,157 @@
+        auto rng_event_1 = mkl_rng::generate(mkl_rng::uniform<DataType>(5.0, 50.0), engine, num_options, h_StockPrice_ptr);
+        auto rng_event_2 = mkl_rng::generate(mkl_rng::uniform<DataType>(10.0, 25.0), engine, num_options, h_OptionStrike_ptr);
+        auto rng_event_3 = mkl_rng::generate(mkl_rng::uniform<DataType>(1.0, 5.0), engine, num_options, h_OptionYears_ptr);
+
+        std::size_t n_states = global_size;
+        using EngineType =
+#if USE_MCG59
+            mkl_rng::device::mcg59<VEC_SIZE>;
+#elif USE_MRG
+            mkl_rng::device::mrg32k3a<VEC_SIZE>;
+#else
+            mkl_rng::device::philox4x32x10<VEC_SIZE>;
+#endif
+
+        // initialization needs only on first step
+        auto deleter = [my_queue](auto* ptr) {sycl::free(ptr, my_queue);};
+        auto rng_states_uptr = std::unique_ptr<EngineType, decltype(deleter)>(sycl::malloc_device<EngineType>(n_states, my_queue), deleter);
+        auto* rng_states = rng_states_uptr.get();
+
+        my_queue.parallel_for(
+            sycl::range<1>(n_states),
+            std::vector<sycl::event>{rng_event_1, rng_event_2, rng_event_3},
+            [=](sycl::item<1> idx) {
+                auto id = idx[0];
+#if USE_MRG
+                constexpr std::uint32_t seed = 12345u;
+                rng_states[id] = EngineType({ seed, seed, seed, seed, seed, seed }, { 0, (4096 * id) });
+#else
+                rng_states[id] = EngineType(rand_seed, id * ITEMS_PER_WORK_ITEM * VEC_SIZE * block_n);
+#endif
+        })
+        .wait_and_throw();
+
+        // main cycle
+        for (int i = 0; i < num_iterations; i++)
+        {
+            tt.start();
+
+            my_queue.parallel_for(
+                sycl::nd_range<1>({global_size}, {local_size}),
+                [=](sycl::nd_item<1> item)
+                {
+                    auto local_state = rng_states[item.get_global_id()];
+
+                    for(std::size_t i = 0; i < ITEMS_PER_WORK_ITEM; ++i)
+                    {
+                        const std::size_t i_options = item.get_group_linear_id() * ITEMS_PER_WORK_ITEM + i;
+                        const DataType VBySqrtT = VLog2E * sycl::sqrt(h_OptionYears_ptr[i_options]);
+                        const DataType MuByT = MuLog2E * h_OptionYears_ptr[i_options];
+                        const DataType Y = h_StockPrice_ptr[i_options];
+                        const DataType Z = h_OptionStrike_ptr[i_options];
+                        DataType v0 = 0, v1 = 0;
+
+                        mkl_rng::device::gaussian<DataType> distr(MuByT, VBySqrtT);
+
+                        for (int block = 0; block < block_n; ++block)
+                        {
+                            auto rng_val_vec = mkl_rng::device::generate(distr, local_state);
+                            for (int lane = 0; lane < VEC_SIZE; ++lane)
+                            {
+                                DataType rng_val = Y * sycl::exp2(rng_val_vec[lane]) - Z;

```

Comment:
Does it make sense to move operations from 121 line above the loop?

`rng_val_vec = Y * sycl::exp2(rng_val_vec);`

Reply:
Yes.